### PR TITLE
Make `Notifications` poll `ProtocolController`s

### PR DESCRIPTION
**Pull request by dmitry-markin**
This PR restores the practice of polling slot allocation logic by `Notifications` and reporting peer connection events via sync calls instead of async channels in order to make `ProtocolController` less susceptible to state inconsistency issues.

### CODE REVIEW

File: client/network/src/peerset.rs [hunk 1]
```diff
@@ -329,15 +312,17 @@ impl Stream for Peerset {
...
-		if let Poll::Ready(msg) = self.from_controllers.poll_next_unpin(cx) {
-			if let Some(msg) = msg {
-				return Poll::Ready(Some(msg))
-			} else {
-				debug!(
-					target: LOG_TARGET,
-					"All `ProtocolController`s have terminated, terminating `Peerset`."
-				);
-				return Poll::Ready(None)
+		for protocol_controller in self.protocol_controllers.iter_mut() {
```
[the remark this answers is not in the document]
Comment (dmitry-markin) [reply client/network/src/peerset.rs]: Addressed in https://github.com/paritytech/substrate/pull/14509/commits/81c87053f4fa5037580ec6cc4f8f812c82059db1.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3139449

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3139669

**Comment by dmitry-markin** [comment]
Closing as we don't plan to merge this now. We might want to revisit the slot allocation implementation in the future though.

**Pull request closed without merging**